psutils.1
.TH PSUTILS 1 "PSUtils"
.SH NAME
PSUtils \- PDF and PostScript utilities
.SH DESCRIPTION
PSUtils is a set of utilities for manipulating PDF and PostScript
documents. PostScript files should follow the Adobe Document Structuring
Conventions, but this is not checked.
.SH UNITS OF LENGTH
PSUtils utilities accept lengths in various units:
.IP \(bu
.B pt
(PostScript points, 72 points per inch)
.IP \(bu
.B mm
(millimetres)
.IP \(bu
.B cm
(centimetres)
.IP \(bu
.B in
(inches, 1 inch is 25.4mm)
.PP
Write the length as a number directly followed by the unit, for example,
.B 4.5mm
or
.BR 72pt .
If no unit is given, PostScript points are assumed.
.SH AUTHOR
Written by Angus J. C. Duggan.
.SH "SEE ALSO"
.BR psbook (1),
.BR psselect (1),
.BR pstops (1),
.BR epsffit (1),
.BR psnup (1),
.BR psresize (1),
.BR psjoin (1),
.BR extractres (1),
.BR includeres (1)
.SH TRADEMARKS
.B PostScript
is a trademark of Adobe Systems Incorporated.


debian/changelog
psutils (2.0-1) unstable; urgency=low

  * New upstream release.

 -- Reuben Thomas <rrt@sc3d.org>  Sat, 25 Jan 2020 13:29:55 +0000

psutils (1.17.dfsg-4) unstable; urgency=medium

  [ Antony Pavlov ]
  * psmerge: Use ps2write.  (pswrite is gone from gs.)
    Closes:#797528.

  [ Bjarni Ingi Gislason ]
  * psnup(1): Fix manpage syntax error.  Closes:#663258.
  * pstops(1): Remove unneeded backslash.  Closes:#726468.

  [ Helmut Grohne <helmut@subdivi.de> ]
  * Fix FTCBFS: Pass CC as a cmdline option to make.  Closes:#838205.

  [ Reuben Thomas ]
  * psnup(1): Remove a confusing and unnecessary qualification
    about units.  Apropos of #695179.

  [ Ian Jackson ]
  * Make myself the maintainer.  Closes:#777699.
  * Fix the upstream URL to refer to Reuben Thomas's github.
    Closes:#752161.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Sat, 21 Jan 2017 19:39:50 +0000

psutils (1.17.dfsg-3) unstable; urgency=medium

  * Orphan package

 -- Jay Berkenbilt <qjb@debian.org>  Fri, 26 Aug 2016 21:59:27 -0400

psutils (1.17.dfsg-2) unstable; urgency=medium

  * Replace hardening with dpkg-buildflags
  * Update standards to 3.9.5

 -- Jay Berkenbilt <qjb@debian.org>  Sat, 26 Jul 2014 19:09:32 -0400

psutils (1.17.dfsg-1) unstable; urgency=low

  * Repackage without files with unclear copyright.  See copyright file
    and README.source for details.  (Closes: #682708)

 -- Jay Berkenbilt <qjb@debian.org>  Sat, 25 Aug 2012 06:10:34 -0400

psutils (1.17-32) unstable; urgency=low

  * Hardening, update standards version

 -- Jay Berkenbilt <qjb@debian.org>  Fri, 29 Jun 2012 21:23:03 -0400

psutils (1.17-31) unstable; urgency=low

  * Mark psutils Multi-Arch: foreign.  (Closes: #649693)

 -- Jay Berkenbilt <qjb@debian.org>  Mon, 28 Nov 2011 10:33:39 -0500

psutils (1.17-30) unstable; urgency=low

  * debian/patches/bboxesandsizes.patch: add page headers to output
    according to parameters (LP #396025).  Thanks Michael Braun.
    (Closes: #632182)
  * Update standards to 3.9.2

 -- Jay Berkenbilt <qjb@debian.org>  Sat, 03 Sep 2011 10:40:07 -0400

psutils (1.17-29) unstable; urgency=low

  * Update large file patch to correct regression caused on kfreebsd-i386.
    Thanks Guillem Jover. (Closes: #617872)
  * Remove -Werror from compilation flags as per correct suggestion from
    Guillem Jover.

 -- Jay Berkenbilt <qjb@debian.org>  Sat, 02 Apr 2011 18:54:02 -0400

psutils (1.17-28) unstable; urgency=low

  * Change source format to '3.0 (quilt)'
  * Remove README.source
  * Update standards to 3.9.1
  * Accept patch from daniele giacomini to allow psutils to work on files
    larger than 2 GiB.  (Closes: #595239)
  * Accept psutils-flip.patch from Peter Korsgaard.  This patch has been
    in SUSE for some time.  (Closes: #534455)
  * Add public domain psjoin script.  (Closes: #253866)

 -- Jay Berkenbilt <qjb@debian.org>  Sun, 17 Oct 2010 12:39:12 -0400

psutils (1.17-27) unstable; urgency=low

  * Housekeeping: add README.source, convert patches to quilt, add
    homepage to control file, update standards version to 3.8.3.

 -- Jay Berkenbilt <qjb@debian.org>  Wed, 19 Aug 2009 19:17:51 -0400

psutils (1.17-26) unstable; urgency=low

  * Change Recommends on gs | gs-esp to one on ghostscript.

 -- Jay Berkenbilt <qjb@debian.org>  Mon, 30 Jun 2008 16:59:26 -0400

psutils (1.17-25) unstable; urgency=low

  * Update standards version to 3.7.3.  No changes required.
  * Fix error in psmerge manual page.  (Closes: #464785)
  * Include patch from Bernhard R. Link to remove paper size information
    and bounding boxes from psnup and psresize.  This removes bugs
    resulting in having usually incorrect information in files when the
    paper size has changed, reasoning that having no information is better
    than having incorrect information.  (Closes: #368331, #434500)
  * Honor DEB_BUILD_OPTIONS=nostrip by not explicitly calling install -s
    to strip executables and instead letting dh_strip do it for us.
    (Closes: #437832)

 -- Jay Berkenbilt <qjb@debian.org>  Sun, 08 Jun 2008 11:16:42 -0400

psutils (1.17-24) unstable; urgency=low

  * Yet another fix to the pstops manual page involving handling of quotes
    around examples. (Closes: #392737)

 -- Jay Berkenbilt <qjb@debian.org>  Sat, 18 Nov 2006 13:16:01 -0500

psutils (1.17-23) unstable; urgency=low

  * Correction to quotation changes in manual pages to work better in
    multiple locales.
  * Add watch file
  * Update URL in copyright file

 -- Jay Berkenbilt <qjb@debian.org>  Fri, 23 Jun 2006 14:47:32 -0400

psutils (1.17-22) unstable; urgency=low

  * From Ian Jackson <iwj@ubuntu.com>: Make psnup (pstops, etc. etc.) cope
    with output from Gnome printing system, which does `/showpage load',
    thus defeating attempts to redefine showpage from virtual page to
    virtual page.  Instead, we redefine showpage to a procedure which
    looks at PStoPSenablepage, a new variable, which is what changes from
    page to page.  Malone #28311.  (Closes: #360606)
  * List binaries in description.  I had removed these but have been
    convinced that they belong after all.  (Closes: #357560)
  * Add quotation marks to examples with parentheses in pstops manual
    page.  Thanks Danai SAE-HAN.  (Closes: #363092)
  * Updated standards version to 3.7.2.  No changes required.

 -- Jay Berkenbilt <qjb@debian.org>  Mon, 29 May 2006 10:40:58 -0400

psutils (1.17-21) unstable; urgency=low

  * Fix packaging bug that prevented /etc/papersize from being inspected.
    (Closes: #339506)

 -- Jay Berkenbilt <qjb@debian.org>  Sat, 19 Nov 2005 11:57:43 -0500

psutils (1.17-20) unstable; urgency=low

  * New maintainer.  Thanks Rob!  (Closes: #329425)
  * Bring package to current Standards-Version; make other miscellaneous
    fixes to control file
  * No longer Provide postscript-previewer
  * Modify psmerge to use gs if available; make this package Recommmend
    (rather than Suggest) some version of gs (Closes: #159888, #35328)

psutils.py
# PSUtils utility library
# Copyright (c) Reuben Thomas 2023.
# Released under the GPL version 3, or (at your option) any later version.

from __future__ import annotations

import os
import sys
import argparse
import shutil
import tempfile
import subprocess
import re
import warnings
from warnings import warn
from typing import (
    Any, List, Optional, Union, Type, NoReturn,
)

import __main__

# Help output
# Adapted from https://stackoverflow.com/questions/23936145/python-argparse-help-message-disable-metavar-for-short-options
class HelpFormatter(argparse.RawTextHelpFormatter):
    def _format_action_invocation(self, action):
        if not action.option_strings:
            metavar, = self._metavar_formatter(action, action.dest)(1)
            return metavar
        else:
            parts = []
            if action.nargs == 0:
                # Option takes no argument, output: -s, --long
                parts.extend(action.option_strings)
            else:
                # Option takes an argument, output: -s, --long ARGUMENT
                default = action.dest.upper()
                args_string = self._format_args(action, default)
                for option_string in action.option_strings:
                    parts.append(option_string)
                parts[-1] += f' {args_string}'
            # Add space at start of format string if there is no short option
            if len(action.option_strings) > 0 and action.option_strings[0][1] == '-':
                parts[-1] = '    ' + parts[-1]
            return ', '.join(parts)

# Error messages
def simple_warning( # pylint: disable=too-many-arguments
        message: Union[Warning, str],
        category: Type[Warning], # pylint: disable=unused-argument
        filename: str, # pylint: disable=unused-argument
        lineno: int, # pylint: disable=unused-argument
        file: Optional[TextIO] = sys.stderr, # pylint: disable=redefined-outer-name
        line: Optional[str] = None # pylint: disable=unused-argument
) -> None:
    print(f'\n{__main__.parser.prog}: {message}', file=file or sys.stderr)
warnings.showwarning = simple_warning

def die(msg: str, code: Optional[int] = 1) -> NoReturn:
    warn(msg)
    sys.exit(code)

# Adapted from https://github.com/python/cpython/blob/main/Lib/test/test_strtod.py
strtod_parser = re.compile(r'''    # A numeric string consists of:
    [-+]?          # an optional sign, followed by
    (?=\d|\.\d)    # a number with at least one digit
    \d*            # having a (possibly empty) integer part
    (?:\.(\d*))?   # followed by an optional fractional part
    (?:E[-+]?\d+)? # and an optional exponent
''', re.VERBOSE | re.IGNORECASE).match

def strtod(s: str):
    m = strtod_parser(s)
    if m is None:
        raise ValueError('invalid numeric string')
    return float(m[0]), m.end()

# Argument parsers
def singledimen(s: str, width: Optional[float] = None, height: Optional[float] = None):
    num, unparsed = strtod(s)
    s = s[unparsed:]

    if s.startswith('pt'):
        pass
    elif s.startswith('in'):
        num *= 72
    elif s.startswith('cm'):
        num *= 28.346456692913385211
    elif s.startswith('mm'):
        num *= 2.8346456692913385211
    elif s.startswith('w'):
        if width is None:
            die('paper size not set')
        num *= width
    elif s.startswith('h'):
        if height is None:
            die('paper size not set')
        num *= height
    elif s != '':
        die(f"bad dimension `{s}'")

    return num

# Get the size of the given paper, or the default paper if no argument given.
def paper(cmd: List[str], silent: bool):
    cmd.insert(0, 'paper')
    try:
        out = subprocess.check_output(cmd, stderr=subprocess.DEVNULL if silent else None, text=True)
        return out.rstrip()
    except subprocess.CalledProcessError:
        return None
    except:
        die("could not run `paper' command")

def paper_size(paper_name: Optional[str] = None):
    if paper_name is None:
        paper_name = paper([])
    dimensions = paper(['--unit=pt', paper_name], True)
    if dimensions is None:
        return None, None
    m = re.search(' ([.0-9]+)x([.0-9]+) pt$', dimensions)
    w, h = float(m[1]), float(m[2])
    return int(w + 0.5), int(h + 0.5) # round dimensions to nearest point

def parsepaper(paper: str):
    try:
        (width, height) = paper_size(paper)
        if width is None:
            [width, height] = paper.split('x')
            if width and height:
                width, height = singledimen(width), singledimen(height)
        return width, height
    except:
        die(f"paper size '{paper}' unknown")

def parse_input_paper(s: str):
  __main__.iwidth, __main__.iheight = parsepaper(s)

def parse_output_paper(s: str):
  __main__.width, __main__.height = parsepaper(s)

def parsedimen(s: str):
    return singledimen(s, __main__.width, __main__.height)

def parsedraw(s: str):
    return parsedimen(s or '1')

# Return comment keyword and value if `line' is a DSC comment
def comment(line: str) -> Optional[(str, str)]:
    m = re.match(r'%%(\S+)\s+?(.*\S?)\s*$', line)
    if m:
        return m[1], m[2]

# Build array of pointers to start/end of pages
def parse_file(infile: IO, explicit_output_paper: bool = False):
    nesting = 0
    psinfo = {
        'headerpos': 0,
        'pagescmt': 0,
        'endsetup': 0,
        'beginprocset': 0,          # start and end of pstops procset
        'endprocset': 0,
        'pages': None,
        'sizeheaders': [],
        'pageptr': [],
    }
    infile.seek(0)
    record, next_record, buffer = 0, 0, None
    for buffer in infile:
        next_record += len(buffer)
        if buffer.startswith('%%'):
            keyword, value = comment(buffer)
            if keyword is not None:
                if nesting == 0 and keyword == 'Page:':
                    psinfo['pageptr'].append(record)
                elif psinfo['headerpos'] == 0 and explicit_output_paper and \
                    keyword in ['BoundingBox:', 'HiResBoundingBox:', 'DocumentPaperSizes:', 'DocumentMedia:']:
                    # FIXME: read input paper size (from DocumentMedia comment?) if not
                    # set on command line.
                    psinfo['sizeheaders'].append(record)
                elif psinfo['headerpos'] == 0 and keyword == 'Pages:':
                    psinfo['pagescmt'] = record
                elif psinfo['headerpos'] == 0 and keyword == 'EndComments':
                    psinfo['headerpos'] = next_record
                elif keyword in ['BeginDocument:', 'BeginBinary:', 'BeginFile:']:
                    nesting += 1
                elif keyword in ['EndDocument', 'EndBinary', 'EndFile']:
                    nesting -= 1
                elif nesting == 0 and keyword == 'EndSetup':
                    psinfo['endsetup'] = record
                elif nesting == 0 and keyword == 'BeginProlog':
                    psinfo['headerpos'] = next_record
                elif nesting == 0 and buffer == '%%BeginProcSet: PStoPS':
                    psinfo['beginprocset'] = record
                elif psinfo['beginprocset'] is not None and \
                     psinfo['endprocset'] is None and keyword == 'EndProcSet':
                    psinfo['endprocset'] = next_record
                elif nesting == 0 and keyword in ['Trailer', 'EOF']:
                    break
        elif psinfo['headerpos'] == 0:
            psinfo['headerpos'] = record
        record = next_record
    psinfo['pages'] = len(psinfo['pageptr'])
    psinfo['pageptr'].append(record)
    if psinfo['endsetup'] == 0 or psinfo['endsetup'] > psinfo['pageptr'][0]:
        psinfo['endsetup'] = psinfo['pageptr'][0]
    return psinfo

# Set up input and output files
def setup_input_and_output(infile_name: str, outfile_name: str, make_seekable: bool = False) -> (IO, IO):
    if infile_name is not None:
        try:
            infile = open(infile_name)
        except:
            die(f'cannot open input file {infile_name}')
    else:
        infile = sys.stdin
    infile.reconfigure(newline=None)
    if make_seekable:
        infile = seekable(infile)
    if infile is None:
        die('cannot make input seekable')

    if outfile_name is not None:
        try:
            outfile = open(outfile_name, 'w')
        except:
            die(f'cannot open output file {outfile_name}')
    else:
        outfile = sys.stdout
    outfile.reconfigure(newline=None)

    return infile, outfile

# Make a file handle seekable, using a temporary file if necessary
def seekable(fp: IO) -> Optional[IO]:
    if fp.seekable():
        return fp

    try:
        ft = tempfile.TemporaryFile()
        shutil.copyfileobj(fp, ft)
        ft.seek(0)
        return ft
    except IOError:
        return None

# Resource extensions
def extn(ext):
    exts = {'font': '.pfa', 'file': '.ps', 'procset': '.ps',
            'pattern': '.pat', 'form': '.frm', 'encoding': '.enc'}
    return exts.get(ext, '')

# Resource filename
def filename(*components): # make filename for resource in 'components'
    name = ''
    for c in components: # sanitise name
        c = re.sub(r'[!()\$\#*&\\\|\`\'\"\~\{\}\[\]\<\>\?]', '', c)
        name += c
    name = os.path.basename(name) # drop directories
    if name == '':
        die(f'filename not found for resource {" ".join(components)}', 2)
    return name
